sys/arm64/arm64/copyinout.S

```diff
@@ -91,28 +112,48 @@ END(copyin)
 /*
  * Copies a string from a user to kernel address
  *
- * int copyinstr(const void *udaddr, void *kaddr, size_t len, size_t *done)
+ * int copyinstr(const void * __capability udaddr, void *kaddr, size_t len,
+ *         size_t *done)
  */
 ENTRY(copyinstr)
 	mov	x5, xzr		/* count = 0 */
 	mov	w4, #1		/* If zero return faulure */
 	cbz	x2, 3f		/* If len == 0 then skip loop */
 
-	adr	x6, copyio_fault /* Get the handler address */
+#if __has_feature(capabilities)
+	cvtd	c1, x1			/* Store through a capability */
+	scbnds	c1, c1, x2		/* Set the bounds */
+#endif
+
+	adr	x6, copyio_c64_fault /* Get the handler address */
```

Comment:
Hm, I don't like that we call it C64 even in non-CHERI kernels. But also, shouldn't it "just work" by keying off the low bit of the address in pcb_onfault (which will need a copy of the code used for signal handlers)?

Reply:
Also it's a little dangerous to do this _before_ entering C64... you'd hope to not get a fault before then, but things will go off the rails if you do

Reply:
There shouldn't be anywhere we could fault between setting the fault handler and entering C64 mode. The only instruction is the `ldr`, however as this is a pseudo instruction the assembler will use a `mov` so won't cause a trap.

It's possible PCC is set so the bounds are incorrect & we trap, but that should never happen, and if it does would indicate larger problems. Interrupts are also OK as they should work at any point in this code without changing it's behaviour.

Reply:
But if you flip the order then it'll be safe and panic on an unexpected trap.

Reply:
If I flip the order the `SET_FAULT_HANDLER` macro won't work. The current order is safe as we need there to be specific types of aborts for the fault handler to be called that the kernel is careful to not cause.

Reply:
It's just loads and stores, shouldn't the alternate mode ones work in C64?

Reply:
There is no alt mode `ldtrb`

Reply:
Sorry, mis-read what you wrote.